docs/LoongArch-ELF-ABI-EN.adoc

```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
+
+=== Argument passing
+
+Generally speaking, FARs are only used to pass floating-point arguments, GARs are used to pass non floating-point arguments and floating-point arguments when no FAR is available(long double type is also passed in a pair of GARs) and the reference.
+
+Arguments passed by reference may be modified by the callee.
+
+==== Scalar
+
+There are three cases:
+
+. 0 < WOA ≤ GRLEN
+
+.. Argument is passed in a single AR, or on the stack by value if none is available.
+
+... If the argument is floating-point type, the argument is passed in FAR.
+if no FAR is available, it’s passed in GAR. If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, floating-point types narrower than GRLEN bits are widened to GRLEN bits, with the upper bits undefined.
+
+... If the argument is integer or pointer type, the argument is passed in GAR.
+If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, the unsigned integer scalars narrower than GRLEN bits are zero-extended to GRLEN bits, and the signed integer scalars are sign-extended.
+
+. GRLEN < WOA ≤ 2 × GRLEN
+
+.. The argument is passed in a pair of GAR, with the low-order GRLEN bits in the lower-numbered register and the high-order GRLEN bits in the higher-numbered register.
+If exactly one register is available, the low-order GRLEN bits are passed in the register and the high-order GRLEN bits are passed on the stack.
+If no GAR is available, it’s passed on the stack.
+
+. WOA > 2 × GRLEN
+
+.. Scalars are passed by reference and are replaced in the argument list with the address.
+
+==== Structure
+
+Empty structures are ignored by C compilers which support them as a non-standard extension(same as union arguments and return values).
+Bits unused due to padding, and bits past the end of a structure whose size in bits is not divisible by GRLEN, are undefined.
+And the layout of the structure on the stack is consistent with that in memory.
+
+. 0 < WOA ≤ GRLEN
+
+.. The structure has only fixed-point members.
+If there is an available GAR, the structure is passed through the GAR by value passing; If no GAR is available, it’s passed on the stack.
+
+.. The structure has only floating-point members:
+
+... One floating-point member.
+The argument is passed in a FAR; If no FAR is available, the value is passed in a GAR; if no GAR is available, the value is passed on the stack.
+
+... Two floating-point members.
+The argument is passed in a pair of available FAR, with the low-order float member bits in the lower-numbered FAR and the high-order float member bits in the higher-numbered FAR.
+If the number of available FAR is less than 2, it’s passed in a GAR, and passed on the stack if no GAR is available.
+
+.. The structure has both fixed-point and floating-point members, i.e. the structure has one float member and
```

Comment:
```suggestion
.. The structure has both fixed-point and floating-point members, i.e. the structure has one floating-point member and
```

Reply:
Other then that. When read this line, I initially though that it is half-written. Then I realize that this is trick to simplify shared naming in headers.
I would suggest add indication that this is not full header, for example like this
```
.. The structure has both fixed-point and floating-point members, i.e. the structure has one float member and...

... multiple fixed-point members.

... only one fixed-point member.
```
or other way
```
.. The structure has both fixed-point and floating-point members, i.e. the structure has one float member and

... ...multiple fixed-point members.

... ...only one fixed-point member.
```

`...` serve as visual hint that this is header would be not complete and some continuation lays somewhere down the road.

Reply:
Thanks a lot.